walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendeesRequested = request.getAttendees();
+    int meetingDuration = (int) request.getDuration();
+
+    // check for no attendees
+    if (attendeesRequested.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // check duration of meeting request
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    List<TimeRange> badTimes = getBadTimes(events, attendeesRequested);
+
+    // list to hold all the available times (everything not in badTimes)
+    List<TimeRange> goodTimes = getGoodTimes(badTimes, meetingDuration);
+    return goodTimes;
+  }
+
+  /** Returns a List with all the times that requested attendees have conflicts. */
+  private List<TimeRange> getBadTimes(
+      Collection<Event> events, Collection<String> attendeesRequested) {
+    List<TimeRange> badTimes = new ArrayList<TimeRange>();
+
+    // iterate through all events and add  time to badTimes if attendee is in attendeesRequested
+    for (Event conflictEvent : events) {
+      Set<String> busyAttendees = conflictEvent.getAttendees();
+
+      // loop to see if any attendees in conflictEvent are in attendeesRequested
+      Boolean conflict = false;
+      Iterator<String> eventIter = busyAttendees.iterator();
+      while (!conflict && eventIter.hasNext()) {
+        if (attendeesRequested.contains(eventIter.next())) {
+          conflict = true;
+        }
+      }
+      if (conflict == true) { // there was an attendee with a conflict, so add bad time to List
+        badTimes.add(conflictEvent.getWhen());
+      }
+    }
+
+    // sort badTimes
+    Collections.sort(badTimes, TimeRange.ORDER_BY_START);
+    return badTimes;
+  }
+
+  /**
+   * Returns a List with all the TimeRanges between the badTimes that are greater or equal to the
+   * requested meeting duration.
+   */
+  private List<TimeRange> getGoodTimes(List<TimeRange> badTimes, int meetingDuration) {
+    List<TimeRange> goodTimes = new ArrayList<TimeRange>();
+
+    // beginning of an available period
+    int goodStart = TimeRange.START_OF_DAY;
+
+    // end of an available period
+    int goodEnd = TimeRange.END_OF_DAY;
+
+    for (TimeRange badRange : badTimes) {
+      goodEnd = badRange.start();
+
+      if (goodEnd > goodStart) {
+        TimeRange goodRange = TimeRange.fromStartEnd(goodStart, goodEnd, false);
+        if (goodRange.duration() >= meetingDuration) {
+          goodTimes.add(goodRange);
+        }
+      }
+
+      // only increment goodStart if it is before the next ending period - takes care of nested
+      // events
+      if (goodStart < (goodEnd + badRange.duration())) {
+        goodStart = goodEnd + badRange.duration();
+      }
```

Comment:
Hm, I think this could be clearer; Since goodEnd is badRange.start(), goodEnd + badRange.duration() is just badRange.end(), thus this is just:

if (goodStart < badRange.end()) goodStart = badRange.end()

which can be further reduced to:

goodStart = Math.max(goodStart, badRange.end())

Reply:
thanks! that is much clearer